frame/data-feed/src/lib.rs

```diff
@@ -0,0 +1,704 @@
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Encode, Decode, RuntimeDebug)]
+pub enum DataType {
+	U128(u128),
+	FixedU128(FixedU128),
+}
+
+impl From<u128> for DataType {
+	fn from(n: u128) -> Self {
+		DataType::U128(n)
+	}
+}
+
+impl From<FixedU128> for DataType {
+	fn from(f: FixedU128) -> Self {
+		DataType::FixedU128(f)
+	}
+}
+
+impl DataType {
+	pub fn number_type(&self) -> NumberType {
+		match self {
+			DataType::U128(_) => NumberType::U128,
+			DataType::FixedU128(_) => NumberType::FixedU128,
+		}
+	}
+
+	pub fn into_fixed_u128(self) -> Option<FixedU128> {
+		match self {
+			DataType::FixedU128(f) => Some(f),
+			_ => None,
+		}
+	}
+
+	pub fn from_number_value(val: NumberValue, target_type: NumberType) -> Option<Self> {
+		match target_type {
+			NumberType::U128 => {
+				if val.integer < 0 {
+					return None;
+				}
+				Some(DataType::U128(val.integer as u128))
+			}
+			NumberType::FixedU128 => {
+				if val.integer < 0 {
+					return None;
+				}
+				let decimal_point = -(val.fraction_length as i32) + val.exponent;
+				let (n, d) = if decimal_point >= 0 {
+					// e.g. 1.35 * 10^3
+					// integer part is (1 * 10^2 + 35) * 10^(-2+3)
+					// decimal part is 1
+					let n = (val.integer * 10_i64.pow(val.fraction_length) + val.fraction as i64)
+						* 10_i64.pow(decimal_point as u32);
+					let d = 1;
+					(n, d)
+				} else {
+					// e.g. 1.35 * 10^-2
+					// integer part is (1 * 10^2 + 35)
+					// decimal part is 10^(|-2-2|)
+					let n = val.integer * 10_i64.pow(val.fraction_length) + val.fraction as i64;
+					// let d = decimal_point.abs();
+					let d = 10_i64.pow(decimal_point.abs() as u32);
+					(n, d)
+				};
+				let res = FixedU128::from((n, d));
+				Some(DataType::FixedU128(res))
+			}
+		}
+	}
+}
+
+/// Types of DataType
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Encode, Decode, RuntimeDebug)]
+pub enum NumberType {
+	U128,
+	FixedU128,
+}
+
+impl Default for DataType {
+	fn default() -> Self {
+		DataType::U128(0)
+	}
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Encode, Decode, RuntimeDebug)]
+pub struct Info<BlockNumber> {
+	/// the key of what we want from the json callback
+	key_str: Vec<u8>,
+	number_type: NumberType,
+	operation: Operations,
+	schedule: BlockNumber,
+}
+
+impl<BlockNumber> Info<BlockNumber> {
+	fn key_str(&self) -> &[u8] {
+		&self.key_str
+	}
+
+	fn number_type(&self) -> NumberType {
+		self.number_type
+	}
+}
+
+pub trait Trait: CreateSignedTransaction<Call<Self>> {
+	/// The identifier type for an offchain worker.
+	type AuthorityId: AppCrypto<Self::Public, Self::Signature>;
+	/// Because this pallet emits events, it depends on the runtime's definition of an event.
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+	/// The overarching dispatch call type.
+	type Call: From<Call<Self>>;
+	/// The origin that can schedule an update
+	type DispatchOrigin: EnsureOrigin<Self::Origin, Success = Self::AccountId>;
```

Comment:
Probably a better name could be picked for this origin to be a bit more descriptive

Reply:
fixed